static/app/views/ddm/metricScratchpad.tsx

```diff
@@ -0,0 +1,266 @@
+  default: string | null;
+  scratchpads: Record<string, Scratchpad>;
+};
+
+export function useScratchpads() {
+  const [state, setState] = useLocalStorageState<ScratchpadState>('ddm-scratchpads', {
+    default: null,
+    scratchpads: {},
+  });
+  const [selected, setSelected] = useState<string | null | undefined>(); // scratchpad id
+
+  const select = useCallback(
+    (id: string | null) => {
+      setSelected(id);
+    },
+    [setSelected]
+  );
+
+  const setDefault = useCallback(
+    (id: string | null) => {
+      setState({...state, default: id});
+    },
+    [state, setState]
+  );
+
+  const add = useCallback(
+    (name: string, query: Scratchpad['query']) => {
+      const id = uuid4();
+      const newScratchpads = {...state.scratchpads, [id]: {name, id, query}};
+      setState({...state, scratchpads: newScratchpads});
+      select(id);
+    },
+    [state, setState, select]
+  );
+
+  const update = useCallback(
+    (id: string, query: Scratchpad['query']) => {
+      const oldScratchpad = state.scratchpads[id];
+      const newScratchpads = {...state.scratchpads, [id]: {...oldScratchpad, query}};
+      setState({...state, scratchpads: newScratchpads});
+    },
+    [state, setState]
+  );
+
+  const remove = useCallback(
+    (id: string) => {
+      const newScratchpads = {...state.scratchpads};
+      delete newScratchpads[id];
+      if (state.default === id) {
+        setState({...state, default: null, scratchpads: newScratchpads});
+      } else {
+        setState({...state, scratchpads: newScratchpads});
+      }
+      if (selected === id) {
+        select(null);
+      }
+    },
+    [state, setState, select, selected]
+  );
+
+  return {
+    all: state.scratchpads,
+    default: state.default,
+    selected,
+    add,
+    update,
+    remove,
+    select,
+    setDefault,
+  };
+}
+
+export function ScratchpadSelector() {
+  const scratchpads = useScratchpads();
+  const router = useRouter();
+  const query = router.location.query ?? {};
+  // select the default scratchpad when the component mounts
+  useEffect(() => {
+    if (scratchpads.default && !scratchpads.selected && !query.widgets) {
+      scratchpads.select(scratchpads.default);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // saves all changes to the selected scratchpad to local storage
+  useEffect(() => {
+    if (scratchpads.selected) {
+      scratchpads.update(scratchpads.selected, query);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [query]);
+
+  // changes the query when a scratchpad is selected, clears it when none is selected
+  useEffect(() => {
+    if (scratchpads.selected) {
+      const selectedQuery = scratchpads.all[scratchpads.selected].query;
+      updateQuery(router, selectedQuery);
+    } else if (scratchpads.selected === null) {
+      clearQuery(router);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [scratchpads.selected]);
```

Comment:
I wonder whether the whole router syncing should be in `useScratchpad` 🤔 
The first useEffect could then be replaced by a simple default value for selected, right?

Reply:
Yep, much nicer!